hatchery/src/world.rs

```diff
@@ -82,41 +85,60 @@ impl World {
                 events: vec![],
                 call_stack: CallStack::default(),
                 height: 0,
+                snapshots: BTreeMap::new(),
             },
         )))))
     }
 
-    pub fn persist(&self) -> Result<(), Error> {
+    pub fn persist(&self) -> Result<SnapshotId, Error> {
         let guard = self.0.lock();
         let w = unsafe { &mut *guard.get() };
+        let mut world_snapshot_id = SnapshotId::uninitialized();
+        let mut world_snapshot = WorldSnapshot::new();
         for (module_id, environment) in w.environments.iter() {
             let memory_path = MemoryPath::new(self.memory_path(module_id));
             let snapshot = Snapshot::new(&memory_path)?;
-            environment.inner_mut().set_snapshot_id(snapshot.id());
-            snapshot.save(&memory_path)?;
+            let snapshot_bag = environment.inner_mut().snapshot_bag_mut();
+            world_snapshot_id.add(&snapshot.id());
+            let snapshot_index =
+                snapshot_bag.save_snapshot(&snapshot, &memory_path)?;
+            world_snapshot.add(*module_id, snapshot_index);
         }
-        Ok(())
+        world_snapshot.finalize_id(world_snapshot_id);
+        w.snapshots.insert(world_snapshot_id, world_snapshot);
+        Ok(world_snapshot_id)
     }
-    pub fn restore(&self) -> Result<(), Error> {
+    pub fn restore(&self, world_snapshot_id: &SnapshotId) -> Result<(), Error> {
         let guard = self.0.lock();
         let w = unsafe { &mut *guard.get() };
-        for (module_id, environment) in w.environments.iter() {
-            let memory_path = MemoryPath::new(self.memory_path(module_id));
-            if let Some(snapshot_id) = environment.inner().snapshot_id() {
-                let snapshot = Snapshot::from_id(*snapshot_id, &memory_path)?;
-                snapshot.load(&memory_path)?;
-                println!(
-                    "restored state of module: {:?} from file: {:?}",
-                    module_id_to_name(*module_id),
-                    snapshot.path()
-                );
-            }
-        }
+        let world_snapshot: &WorldSnapshot =
+            w.snapshots.get(world_snapshot_id).ok_or_else(|| {
+                SnapshotError(String::from("world snapshot id not found"))
+            })?;
+        world_snapshot.restore_snapshots(self)?;
         Ok(())
     }
     pub fn memory_path(&self, module_id: &ModuleId) -> PathBuf {
         self.storage_path().join(module_id_to_name(*module_id))
     }
+    pub fn restore_snapshot_with_index(
```

Comment:
The consumers should not have to know about snapshot indexes. (using snapshot_index in a pub method)